File: src/shapes/rapier_convex_polygon_shape_2d.cpp
#include "rapier_convex_polygon_shape_2d.h"

rapier2d::Handle RapierConvexPolygonShape2D::create_rapier_shape() const {
	ERR_FAIL_COND_V(point_count < 3, rapier2d::invalid_handle());
	rapier2d::Vector *rapier_points = (rapier2d::Vector *)alloca(point_count * sizeof(rapier2d::Vector));
	for (int i = 0; i < point_count; i++) {
		rapier_points[i] = rapier2d::Vector{ (points[i].pos.x), (points[i].pos.y) };
	}
	return rapier2d::shape_create_convex_polyline(rapier_points, point_count);
}

void RapierConvexPolygonShape2D::set_data(const Variant &p_data) {
#ifdef REAL_T_IS_DOUBLE
	ERR_FAIL_COND(p_data.get_type() != Variant::PACKED_VECTOR2_ARRAY && p_data.get_type() != Variant::PACKED_FLOAT64_ARRAY);
#else
	ERR_FAIL_COND(p_data.get_type() != Variant::PACKED_VECTOR2_ARRAY && p_data.get_type() != Variant::PACKED_FLOAT32_ARRAY);
#endif

	if (points) {
		memdelete_arr(points);
	}
	points = nullptr;
	point_count = 0;

	if (p_data.get_type() == Variant::PACKED_VECTOR2_ARRAY) {
		PackedVector2Array arr = p_data;
		ERR_FAIL_COND(arr.size() == 0);
		point_count = arr.size();
		points = memnew_arr(Point, point_count);
		const Vector2 *r = arr.ptr();

		for (int i = 0; i < point_count; i++) {
			points[i].pos = r[i];
		}

		for (int i = 0; i < point_count; i++) {
			Vector2 p = points[i].pos;
			Vector2 pn = points[(i + 1) % point_count].pos;
			points[i].normal = (pn - p).orthogonal().normalized();
		}
	} else {
#ifdef REAL_T_IS_DOUBLE
		PackedFloat64Array dvr = p_data;
#else
		PackedFloat32Array dvr = p_data;
#endif
		point_count = dvr.size() / 4;
		ERR_FAIL_COND(point_count == 0);

		points = memnew_arr(Point, point_count);
		const real_t *r = dvr.ptr();

		for (int i = 0; i < point_count; i++) {
			int idx = i << 2;
			points[i].pos.x = r[idx + 0];
			points[i].pos.y = r[idx + 1];
			points[i].normal.x = r[idx + 2];
			points[i].normal.y = r[idx + 3];
		}
	}

	ERR_FAIL_COND(point_count == 0);
	Rect2 aabb;
	aabb.position = points[0].pos;
	for (int i = 1; i < point_count; i++) {
		aabb.expand_to(points[i].pos);
	}

	configure(aabb);
}

Variant RapierConvexPolygonShape2D::get_data() const {
	PackedVector2Array dvr;

	dvr.resize(point_count);

	for (int i = 0; i < point_count; i++) {
		dvr.set(i, points[i].pos);
	}

	return dvr;
}

real_t RapierConvexPolygonShape2D::get_moment_of_inertia(real_t p_mass, const Size2 &p_scale) const {
	ERR_FAIL_COND_V_MSG(point_count == 0, 0, "Convex polygon shape has no points.");
	Rect2 aabb_new;
	aabb_new.position = points[0].pos * p_scale;
	for (int i = 0; i < point_count; i++) {
		aabb_new.expand_to(points[i].pos * p_scale);
	}

	return p_mass * aabb_new.size.dot(aabb_new.size) / 12.0;
}

RapierConvexPolygonShape2D::~RapierConvexPolygonShape2D() {
	if (points) {
		memdelete_arr(points);
	}
}

File: src/shapes/rapier_convex_polygon_shape_2d.rs
use godot::classes::physics_server_2d::ShapeType;
use godot::prelude::*;

use crate::bodies::vector_normalized;
use crate::rapier_wrapper::prelude::*;
use crate::shapes::rapier_shape::IRapierShape;
use crate::shapes::rapier_shape::RapierShapeBase;
use crate::Vector;
pub struct RapierConvexPolygonShape2D {
    points: Vec<Point>,
    pub base: RapierShapeBase,
}
#[derive(Clone, Copy)]
struct Point {
    pos: Vector2,
    normal: Vector2,
}
impl RapierConvexPolygonShape2D {
    pub fn new(rid: Rid) -> Self {
        Self {
            points: Vec::new(),
            base: RapierShapeBase::new(rid),
        }
    }
}
impl IRapierShape for RapierConvexPolygonShape2D {
    fn get_base(&self) -> &RapierShapeBase {
        &self.base
    }

    fn get_mut_base(&mut self) -> &mut RapierShapeBase {
        &mut self.base
    }

    fn get_type(&self) -> ShapeType {
        ShapeType::CONVEX_POLYGON
    }

    fn get_moment_of_inertia(&self, mass: f32, scale: Vector) -> f32 {
        if self.points.len() < 3 {
            return 0.0;
        }
        let mut aabb_new = Rect2::new(Vector2::ZERO, Vector2::ZERO);
        for point in self.points.iter() {
            aabb_new = aabb_new.expand(point.pos * scale);
        }
        mass * aabb_new.size.dot(aabb_new.size) / 12.0
    }

    fn allows_one_way_collision(&self) -> bool {
        true
    }

    fn create_rapier_shape(&mut self) -> Handle {
        if self.points.len() >= 3 {
            let mut rapier_points = Vec::with_capacity(self.points.len());
            for point in self.points.iter() {
                rapier_points.push(vector_to_rapier(point.pos));
            }
            shape_create_convex_polyline(rapier_points)
        }
        else {
            godot_error!("ConvexPolygon2D must have at least three point");
            invalid_handle()
        }
    }

    fn set_data(&mut self, data: Variant) {
        match data.get_type() {
            VariantType::PACKED_VECTOR2_ARRAY => {
                let arr: PackedVector2Array = data.to();
                let size = arr.len();
                if size <= 0 {
                    return;
                }
                self.points = Vec::with_capacity(size);
                for i in 0..size {
                    self.points.push(Point {
                        pos: arr[i],
                        normal: Vector2::ZERO,
                    });
                }
                for i in 0..size {
                    let p = self.points[i].pos;
                    let pn = self.points[(i + 1) % size].pos;
                    self.points[i].normal = vector_normalized((pn - p).orthogonal());
                }
            }
            VariantType::PACKED_FLOAT32_ARRAY => {
                let arr: PackedFloat32Array = data.to();
                let size = arr.len() / 4;
                if size <= 0 {
                    return;
                }
                self.points = Vec::with_capacity(size);
                for i in 0..size {
                    let idx = i << 2;
                    self.points.push(Point {
                        pos: Vector2::new(arr[idx], arr[idx + 1]),
                        normal: Vector2::new(arr[idx + 2], arr[idx + 3]),
                    });
                }
            }
            VariantType::PACKED_FLOAT64_ARRAY => {
                let arr: PackedFloat64Array = data.to();
                let size = arr.len() / 4;
                if size <= 0 {
                    return;
                }
                self.points = Vec::with_capacity(size);
                for i in 0..size {
                    let idx = i << 2;
                    self.points.push(Point {
                        pos: Vector2::new(arr[idx] as f32, arr[idx + 1] as f32),
                        normal: Vector2::new(arr[idx + 1] as f32, arr[idx + 3] as f32),
                    });
                }
            }
            _ => {
                godot_error!("Invalid shape data");
                return;
            }
        }
        if self.points.len() < 3 {
            godot_error!("ConvexPolygon2D must have at least three point");
            return;
        }
        let mut aabb = Rect2::new(Vector2::ZERO, Vector2::ZERO);
        for point in self.points.iter() {
            aabb = aabb.expand(point.pos);
        }
        let handle = self.create_rapier_shape();
        self.base.set_handle(handle, aabb);
    }

    fn get_data(&self) -> Variant {
        let mut arr = PackedVector2Array::new();
        for point in self.points.iter() {
            arr.push(point.pos);
        }
        arr.to_variant()
    }

    fn get_handle(&mut self) -> Handle {
        self.base.get_handle()
    }
}
